### ISSUE TITLE
(aws-codepipeline): Multiple Cross Account Pipelines Using Same Support Stack

### ISSUE THREAD

**Issue by MrSwayne** [aws/aws-cdk#18433]
### What is the problem?

When utilising CodePipeline and cross account actions, a support stack is created in the target account under the name `${pipelineName}-support-${targetAccountNumber}`. This stack contains the roles needed to create/update cloudformation stacks in the target account.

For our use case we have multiple pipelines which are under the same cdk app (One is the production pipeline, then we have a dev pipeline so that we can test changes to the build/deploy actions before updating the main pipeline. The dev pipeline reuses the same constructs as the production pipeline, but only deploys to our dev environment). 

It seems that when using multiple pipelines in the same app, only the support stack belonging to the first pipeline construct will be used for all pipelines.

This means that deploying our development pipeline deletes all the roles for our production pipeline in the target account. This is problematic as we use these deployment roles in the target accounts KMS key policies. This deletion of the roles also causes the other pipeline to fail as the roles no longer exist for it to assume.

### Reproduction Steps

create a new directory and init cdk
`mkdir crossaccountpipeline && cd crossaccountpipeline`
`cdk init --language typescript`

Update the following files (swap out account numbers with accounts you have access to):

bin/crossaccountpipeline.ts
```
#!/usr/bin/env node
import "source-map-support/register";
import * as cdk from "monocdk";
import { CrossAccountPipelineStack } from "../lib/crossaccountpipeline-stack";

const app = new cdk.App();

const sourceAccount = "407977089617";
const destinationAccount = "735612164109";

const env = {
	region: "us-west-2",
	account: sourceAccount,
};

new CrossAccountPipelineStack(app, "CrossAccountPipelineStack", {
	env,
	destinationAccount,
});

new CrossAccountPipelineStack(app, `CrossAccountPipelineStack-dev`, {
	env,
	destinationAccount,
});

```

lib/crossaccountpipeline-stack.ts

```
import * as cdk from "monocdk";
import * as CodePipeline from "monocdk/aws-codepipeline";
import { CloudFormationCreateUpdateStackAction, EcrSourceAction } from "monocdk/aws-codepipeline-actions";
import * as ECR from "monocdk/aws-ecr";

export interface CrossAccountPipelineStackProps extends cdk.StackProps {
  destinationAccount:string
}

export class CrossAccountPipelineStack extends cdk.Stack {
	constructor(scope: cdk.Construct, id: string, props: CrossAccountPipelineStackProps) {
		super(scope, id, props);

		const pipeline = new CodePipeline.Pipeline(this, `pipeline`, {});
		const sourceOutput = new CodePipeline.Artifact();
		const repository = new ECR.Repository(this, `ecr`, {
			repositoryName: id.toLowerCase(),
		});

		pipeline.addStage({
			stageName: `source`,
			actions: [
				new EcrSourceAction({
					actionName: `ecr-source`,
					repository,
					imageTag: `some-tag`,
					output: sourceOutput,
				}),
			],
		});

		pipeline.addStage({
			stageName: "deploy",

			actions: [
				new CloudFormationCreateUpdateStackAction({
					actionName: `cloudformation-deploy`,
					region: `us-west-2`,
					account: props.destinationAccount,
					stackName: `test-stack`,
					adminPermissions: true,
					templatePath: sourceOutput.atPath("template.yaml"),
				}),
			],
		});
	}
}


```

cdk.json
```
{
	"app": "npx ts-node --prefer-ts-exts bin/crossaccountpipeline.ts",
	"plugin": ["cdk-assume-role-credential-plugin"],

	"watch": {
		"include": ["**"],
		"exclude": [
			"README.md",
			"cdk*.json",
			"**/*.d.ts",
			"**/*.js",
			"tsconfig.json",
			"package*.json",
			"yarn.lock",
			"node_modules",
			"test"
		]
	},
	"context": {
		"@aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId": true,
		"@aws-cdk/core:enableStackNameDuplicates": true,
		"aws-cdk:enableDiffNoFail": true,
		"@aws-cdk/core:stackRelativeExports": true,
		"@aws-cdk/aws-ecr-assets:dockerIgnoreSupport": true,
		"@aws-cdk/aws-secretsmanager:parseOwnedSecretName": true,
		"@aws-cdk/aws-kms:defaultKeyPolicies": true,
		"@aws-cdk/aws-s3:grantWriteWithoutAcl": true,
		"@aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount": true,
		"@aws-cdk/aws-rds:lowercaseDbIdentifier": true,
		"@aws-cdk/aws-efs:defaultEncryptionAtRest": true,
		"@aws-cdk/aws-lambda:recognizeVersionProps": true,
		"@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021": true
	}
}
```

package.json
```
{
  "name": "crossaccountpipeline",
  "version": "0.1.0",
  "bin": {
    "crossaccountpipeline": "bin/crossaccountpipeline.js"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "cdk": "cdk"
  },
  "devDependencies": {
    "@aws-cdk/assertions": "1.135.0",
    "@types/jest": "^26.0.10",
    "@types/node": "10.17.27",
    "jest": "^26.4.2",
    "ts-jest": "^26.2.0",
    "ts-node": "^9.0.0",
    "typescript": "~3.9.7"
  },
  "dependencies": {
    "cdk-assume-role-credential-plugin": "^1.4.0",
    "monocdk": "^1.138.2",
    "source-map-support": "^0.5.16"
  }
}

```

### What did you expect to happen?

after executing `cdk diff CrossAccountPipelineStack`
```
Including dependency stacks: cross-account-support-stack-735612164109
Stack cross-account-support-stack-735612164109 (CrossAccountPipelineStack-support-735612164109)
...
```
after executing `cdk diff CrossAccountPipelineStack-dev`
```
Including dependency stacks: cross-account-support-stack-735612164109
Stack cross-account-support-stack-735612164109 (CrossAccountPipelineStack-dev-support-735612164109)
...
```

### What actually happened?

after executing `cdk diff CrossAccountPipelineStack`
```
Including dependency stacks: cross-account-support-stack-735612164109
Stack cross-account-support-stack-735612164109 (CrossAccountPipelineStack-support-735612164109)
...
```
after executing `cdk diff CrossAccountPipelineStack-dev`
```
Including dependency stacks: cross-account-support-stack-735612164109
Stack cross-account-support-stack-735612164109 (CrossAccountPipelineStack-support-735612164109)
...
```

### CDK CLI Version

1.135.0 (build 3825f59)

### Node.js Version

v14.15.4

### OS

Windows 10

### Language

Typescript

### Language Version

Typescript (4.2.4)

### Other information

I believe the error is in this part of the code

`aws-cdk/packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts`
```
   let targetAccountStack: Stack | undefined = this._crossAccountSupport[targetAccount];
   if (!targetAccountStack) {
      const stackId = `cross-account-support-stack-${targetAccount}`;
      const app = this.supportScope();
      targetAccountStack = app.node.tryFindChild(stackId) as Stack;
      if (!targetAccountStack) {
        const actionRegion = action.actionProperties.resource
          ? action.actionProperties.resource.env.region
          : action.actionProperties.region;
        const pipelineStack = Stack.of(this);
        targetAccountStack = new Stack(app, stackId, {
          stackName: `${pipelineStack.stackName}-support-${targetAccount}`,
          env: {
            account: targetAccount,
            region: actionRegion ?? pipelineStack.region,
          },
        });
      }
      this._crossAccountSupport[targetAccount] = targetAccountStack;
    }
```

Basically, whichever pipeline construct is created first is the support stack that all the other pipelines will also use, as the stackId only changes based on the target account number.

I believe the above stackId should be changed to be based on the pipelineName i.e `${pipelineStack.stackName}-support-${targetAccount}`

**Comment by skinny85** [comment]
Hey @MrSwayne,

thanks for opening the issue. Yes, you're right that the cross-account support Stacks are shared between multiple pipelines in the same application. This was done to save on the number of Stacks that are automatically generated. However, I see how this can be tricky in your case.

We recently had a similar issue about the cross-region support Stacks (https://github.com/aws/aws-cdk/issues/18018), and the PR fixing it (https://github.com/aws/aws-cdk/pull/18043) was super simple - it was just a matter of adding a new setting in the `Pipeline` construct, and changing the ID of the cross-account support Stack if that options is passed, similar to what was done for the cross-region support Stacks.

What do you think about that @MrSwayne? Would you be willing to contribute that change to the CDK? Our "Contributing" guide contains detailed instructions on how to do that: https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md.

Thanks,
Adam
